frame/contracts/src/rent.rs

```diff
@@ -14,102 +14,152 @@
+/// the balance lower than [`subsistence_threshold`].
+fn rent_budget<T: Trait>(
+	balance: &BalanceOf<T>,
+	contract: &AliveContractInfo<T>,
+	subsistence_threshold: &BalanceOf<T>,
+) -> BalanceOf<T> {
+	let rent_allowed_to_charge = *balance - *subsistence_threshold;
+	<BalanceOf<T>>::min(contract.rent_allowance, rent_allowed_to_charge)
+}
 
+/// Consider the case for rent payment of the given account and returns a `Verdict`.
+///
+/// The `current_block_number` must be equal to the current block number. Use `handicap` do
+/// change the reference block number. (See `snitch_contract_should_be_evicted` for more details).
+fn consider_case<T: Trait>(
+	account: &T::AccountId,
+	current_block_number: T::BlockNumber,
+	handicap: T::BlockNumber,
+	contract: &AliveContractInfo<T>,
+) -> Verdict<T> {
 	// How much block has passed since the last deduction for the contract.
 	let blocks_passed = {
 		// Calculate an effective block number, i.e. after adjusting for handicap.
 		let effective_block_number = current_block_number.saturating_sub(handicap);
-		let n = effective_block_number.saturating_sub(contract.deduct_block);
-		if n.is_zero() {
-			// Rent has already been paid
-			return (RentOutcome::Exempted, Some(ContractInfo::Alive(contract)));
-		}
-		n
+		effective_block_number.saturating_sub(contract.deduct_block)
 	};
+	if blocks_passed.is_zero() {
+		// Rent has already been paid
+		return Verdict::Exempt;
+	}
 
 	let balance = T::Currency::free_balance(account);
 
 	// An amount of funds to charge per block for storage taken up by the contract.
-	let fee_per_block = {
-		let free_storage = balance
-			.checked_div(&T::RentDepositOffset::get())
-			.unwrap_or_else(Zero::zero);
-
-		let effective_storage_size =
-			<BalanceOf<T>>::from(contract.storage_size).saturating_sub(free_storage);
-
-		effective_storage_size
-			.checked_mul(&T::RentByteFee::get())
-			.unwrap_or(<BalanceOf<T>>::max_value())
-	};
-
+	let fee_per_block = compute_fee_per_block::<T>(&balance, contract);
 	if fee_per_block.is_zero() {
 		// The rent deposit offset reduced the fee to 0. This means that the contract
 		// gets the rent for free.
-		return (RentOutcome::Exempted, Some(ContractInfo::Alive(contract)));
+		return Verdict::Exempt;
 	}
 
-	// The minimal amount of funds required for a contract not to be evicted.
-	let subsistence_threshold = T::Currency::minimum_balance() + T::TombstoneDeposit::get();
-
+	let subsistence_threshold = subsistence_threshold::<T>();
 	if balance < subsistence_threshold {
 		// The contract cannot afford to leave a tombstone, so remove the contract info altogether.
-		<ContractInfoOf<T>>::remove(account);
-		child::kill_storage(&contract.trie_id, contract.child_trie_unique_id());
-		<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), false));
-		return (RentOutcome::Evicted, None);
+		return Verdict::Kill;
 	}
+	let rent_budget = rent_budget::<T>(&balance, contract, &subsistence_threshold);
```

Comment:
Ah, here we make sure that it's actually the case that `rent_budget` is being called with `balance >= subsistence_threshold`, maybe move the check within the function (and return `Option`) to make that more colocated?

Reply:
ah, yeah, makes sense!